sdk/core/core-http/src/credentials/accessTokenRefresher.ts

```diff
@@ -0,0 +1,54 @@
+import { AccessToken, TokenCredential, GetTokenOptions } from "@azure/core-auth";
+
+/**
+ * Helps the core-http token authentication policies with requesting a new token if we're not currently waiting for a new token.
+ */
+export class AccessTokenRefresher {
+  private promise: Promise<AccessToken | undefined> | undefined;
+  private lastCalled: number | undefined;
+
+  constructor(
+    private credential: TokenCredential,
+    private scopes: string | string[],
+    private requiredMillisecondsBeforeNewRefresh: number
+  ) {}
+
+  private readyForNewRefresh() {
+    // We're only ready for a new refresh if the required milliseconds have passed.
+    return (
+      !this.lastCalled || Date.now() - this.lastCalled > this.requiredMillisecondsBeforeNewRefresh
```

Comment:
One problem with the abstraction `AccessTokenRefresher` being a class separate from the `AccessTokenCache` is that this should really take into account whether there is currently a cached valid token. I'm not sure these concepts should be separated.

Reply:
I believe that your concerns are fixed in this code section:

https://github.com/Azure/azure-sdk-for-js/pull/10085/files#diff-65d593c4170e3a543713c89a60a23baaR100-R122

But I could be missing something.

Here's what's happening in that code section in my words:

- We try to get the cached token.
- If we don't get the cached token:
    - We load it with credential.getToken (same behavior as always)
    - Then we set the cached token.
    - Then we start calling the token refresher.
    - If the refresh promise is null, we ignore it.
- If we got an access token:
    - Try to refresh the token.
    - If the refresh promise is null, we ignore it.
    - If the refresh promise is not null, we wait for it, then update the cached token.

Considering that refreshing the token avoids parallel executions, and only gets to run if it's `readyForNewRefresh` (hasn't been called before, or `requiredMillisecondsBeforeNewRefresh` have elapsed since the last refresh).

I could be missing something here! Please let me know if you're able to spot any misbehavior.

Reply:
So, from a meeting we had, the missing case is that if there's no cached token, we should always refresh.